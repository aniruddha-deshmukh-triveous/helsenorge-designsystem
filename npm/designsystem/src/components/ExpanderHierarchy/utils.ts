export type HeadingTags = 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export const getHeadingTag = (htmlMarkup?: HeadingTags): HeadingTags => {
  switch (htmlMarkup) {
    case 'h2':
      return 'h3';
    case 'h3':
      return 'h4';
    case 'h4':
      return 'h5';
    case 'h5':
      return 'h6';
    case 'h6':
      return 'h6';
    default:
      return 'h2';
  }
};
