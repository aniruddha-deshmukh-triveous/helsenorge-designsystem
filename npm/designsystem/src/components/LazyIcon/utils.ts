export const isServerSide = (): boolean => typeof document === 'undefined';
