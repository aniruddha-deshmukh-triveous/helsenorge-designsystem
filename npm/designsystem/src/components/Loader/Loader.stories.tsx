import React from 'react';
import { withKnobs, select } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import Loader, { Overlay } from './Loader';
import { allPaletteNames } from '../../../.storybook/knobs';
import { withA11y } from '@storybook/addon-a11y';
import Title from '../Title/Title';
import Button from '../Button/Button';

const stories = storiesOf('Loader', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <Loader
    size={select('Size (does not work properly)', ['tiny', 'small', 'medium', 'large'], 'small')}
    color={select('Color', allPaletteNames, 'blueberry')}
  />
));

stories.add('All sizes', () => (
  <div style={{ display: 'grid', gridGap: '2rem' }}>
    <Loader size="tiny" color={select('Color', allPaletteNames, 'blueberry')} />
    <Loader size="small" color={select('Color', allPaletteNames, 'blueberry')} />
    <Loader size="medium" color={select('Color', allPaletteNames, 'blueberry')} />
    <Loader size="large" color={select('Color', allPaletteNames, 'blueberry')} />
  </div>
));

stories.add('Loader is centered', () => (
  <div style={{ width: 300, background: 'white' }}>
    <Loader
      size={select('Size (does not work properly)', ['tiny', 'small', 'medium', 'large'], 'small')}
      color={select('Color', allPaletteNames, 'blueberry')}
      center
    />
  </div>
));

stories.add('Inline', () => (
  <div>
    <Title htmlMarkup="span" appearance="title1">
      {'Søker'}
    </Title>
    <Loader color="black" inline size="tiny" />
  </div>
));

stories.add('Overlay screen', () => (
  <div style={{ width: '100%', height: '200vh' }}>
    <h2>{'Fastlegen din er Arnfinn Nesset ved Orkdal sykehjem'}</h2>
    <Loader
      size={select('Size (does not work properly)', ['tiny', 'small', 'medium', 'large'], 'small')}
      color={select('Color', allPaletteNames, 'black')}
      overlay={Overlay.screen}
    />
  </div>
));

stories.add('Overlay parent', () => (
  <div style={{ width: '100%', height: '100vh' }}>
    <h2>{'Fastlegen din er Arnfinn Nesset ved Orkdal sykehjem'}</h2>
    <div>
      <Title>{'Søknad'}</Title>
      <p style={{ color: 'red' }}>{'Søknad om greier'}</p>
      <Button>{'Lagre'}</Button>
      <Loader
        size={select('Size (does not work properly)', ['tiny', 'small', 'medium', 'large'], 'small')}
        color={select('Color', allPaletteNames, 'black')}
        overlay={Overlay.parent}
      />
    </div>
  </div>
));
