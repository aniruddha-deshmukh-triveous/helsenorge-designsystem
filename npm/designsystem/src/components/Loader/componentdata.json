{
  "props": {
    "color": {
      "defaultValue": null,
      "description": "Sets the color of the loader",
      "name": "color",
      "parent": {
        "fileName": "src/components/Loader/Loader.tsx",
        "name": "LoaderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Loader/Loader.tsx",
          "name": "LoaderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "enum",
        "raw": "\"banana\" | \"blueberry\" | \"cherry\" | \"kiwi\" | \"neutral\" | \"plum\" | \"black\" | \"white\" | undefined",
        "value": [
          {
            "value": "undefined"
          },
          {
            "value": "\"banana\""
          },
          {
            "value": "\"blueberry\""
          },
          {
            "value": "\"cherry\""
          },
          {
            "value": "\"kiwi\""
          },
          {
            "value": "\"neutral\""
          },
          {
            "value": "\"plum\""
          },
          {
            "value": "\"black\""
          },
          {
            "value": "\"white\""
          }
        ]
      }
    },
    "className": {
      "defaultValue": null,
      "description": "Adds custom classes to the element.",
      "name": "className",
      "parent": {
        "fileName": "src/components/Loader/Loader.tsx",
        "name": "LoaderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Loader/Loader.tsx",
          "name": "LoaderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    },
    "size": {
      "defaultValue": null,
      "description": "Changes the size of the loader.",
      "name": "size",
      "parent": {
        "fileName": "src/components/Loader/Loader.tsx",
        "name": "LoaderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Loader/Loader.tsx",
          "name": "LoaderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "enum",
        "raw": "\"tiny\" | \"small\" | \"medium\" | \"large\" | undefined",
        "value": [
          {
            "value": "undefined"
          },
          {
            "value": "\"tiny\""
          },
          {
            "value": "\"small\""
          },
          {
            "value": "\"medium\""
          },
          {
            "value": "\"large\""
          }
        ]
      }
    },
    "testId": {
      "defaultValue": null,
      "description": "Sets the data-testid attribute.",
      "name": "testId",
      "parent": {
        "fileName": "src/components/Loader/Loader.tsx",
        "name": "LoaderProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Loader/Loader.tsx",
          "name": "LoaderProps"
        }
      ],
      "required": false,
      "type": {
        "name": "string"
      }
    }
  }
}
