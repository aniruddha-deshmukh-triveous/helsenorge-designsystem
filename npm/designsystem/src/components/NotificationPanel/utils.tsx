import React, { AriaAttributes } from 'react';
import { palette } from '../../theme/palette';
import Icon from '../Icons';

import InfoSignStroke from '../Icons/InfoSignStroke';
import AlertSignStroke from '../Icons/AlertSignStroke';
import AlertSignFill from '../Icons/AlertSignFill';

export const variantToIconMap = {
  info: <Icon svgIcon={InfoSignStroke} color={palette.kiwi900} hoverColor={palette.kiwi900} />,
  warn: <Icon svgIcon={AlertSignStroke} color="black" hoverColor="black" />,
  alert: <Icon svgIcon={AlertSignFill} color={palette.cherry500} hoverColor={palette.cherry500} />,
  alertLabel: <Icon svgIcon={AlertSignStroke} color={palette.cherry500} hoverColor={palette.cherry500} />,
  crisis: <Icon svgIcon={AlertSignFill} color={palette.banana200} hoverColor={palette.banana200} />,
};

interface AriaLabelAttributesConfig {
  label?: string;
  id?: string;
}

type AriaLabelAttributes = Pick<AriaAttributes, 'aria-label' | 'aria-labelledby'>;

export const getAriaLabelAttributes = (config: AriaLabelAttributesConfig): AriaLabelAttributes | undefined => {
  const { label, id } = config;

  if (id) {
    return {
      'aria-labelledby': id,
    };
  }
  if (label) {
    return {
      'aria-label': label,
    };
  }
};
