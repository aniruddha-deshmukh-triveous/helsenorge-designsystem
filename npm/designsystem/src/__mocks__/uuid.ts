import { vi as jest } from 'vitest';

import * as uuidUtils from '../utils/uuid';

const testId = {
  id: 0,
};

/**
 * Returnerer en unik, forutsigbar id hver gang uuid() kalles
 */
jest.spyOn(uuidUtils, 'uuid').mockImplementation(() => {
  testId.id++;
  return `testid-${testId.id}`;
});
