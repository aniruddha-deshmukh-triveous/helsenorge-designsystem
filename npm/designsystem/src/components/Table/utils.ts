import { Breakpoint } from '../../hooks/useBreakpoint';
import { isTouchDevice } from '../../utils/device';
import { BreakpointConfig, SmallViewportVariant } from './Table';

import styles from './styles.module.scss';

/**
 * Sjekk om det skal brukes CSS for å bestemme responsivt utseende på dette breakpoint
 * @param config Konfigurasjon for responsiv oppførsel
 * @returns true om breakpoint bruker CSS
 */
const configUsesCss = (config: BreakpointConfig): boolean =>
  config.variant === SmallViewportVariant.centeredoverflow || config.variant === SmallViewportVariant.block;

/**
 * Lag klassenavn for CSS-config
 * @param config Konfigurasjon for responsiv oppførsel
 * @returns CSS-klassenavn
 */
const mapConfigToClass = (config: BreakpointConfig): string => styles[`table--${config.variant}-${config.breakpoint}`];

/**
 * Sorter konfigurasjon etter breakpoints, fra største til minste
 * @param a Konfigurasjon for responsiv oppførsel
 * @param b Konfigurasjon for responsiv oppførsel
 * @returns Sortert liste
 */
const sortByBreakpointsDescending = (a: BreakpointConfig, b: BreakpointConfig): number =>
  Breakpoint[a.breakpoint] - Breakpoint[b.breakpoint];

/**
 * Sjekk om en konfigurasjon skal brukes for nåværende breakpoint
 * @param a Konfigurasjon for responsiv oppførsel
 * @param breakpoint Nåværende breakpoint
 * @returns true dersom config skal brukes for breakpointet
 */
const isValidForCurrentBreakpoint = (config: BreakpointConfig, breakpoint: Breakpoint): boolean =>
  Breakpoint[config.breakpoint] >= breakpoint;

/**
 * Finn konfigurasjon for nåværende breakpoint
 * @param config Konfigurasjon for responsiv oppførsel
 * @param breakpoint Nåværende breakpoint
 * @returns Konfigurasjon for responsiv oppførsel
 */
const getConfigForBreakpoint = (config: BreakpointConfig | BreakpointConfig[], breakpoint: Breakpoint): BreakpointConfig | undefined => {
  if (Array.isArray(config)) {
    config.sort(sortByBreakpointsDescending);

    return config.find(x => isValidForCurrentBreakpoint(x, breakpoint));
  } else if (config && isValidForCurrentBreakpoint(config, breakpoint)) {
    return config;
  }
};

/**
 * Finn konfigurasjon for nåværende breakpoint og tabellbredde
 * @param config Konfigurasjon for responsiv oppførsel
 * @param breakpoint Nåværende breakpoint
 * @param tableWidth Bredde på tabell i px
 * @returns Konfigurasjon for responsiv oppførsel
 */
export const getCurrentConfig = (
  config: BreakpointConfig | BreakpointConfig[],
  breakpoint: Breakpoint,
  tableWidth: number,
  windowWidth: number
): BreakpointConfig | undefined => {
  const breakpointConfig = getConfigForBreakpoint(config, breakpoint);
  const canUseHorizontalScroll = isTouchDevice();
  const canUseCenteredOverflow = tableWidth <= windowWidth;

  if (!breakpointConfig) {
    return;
  }

  if (
    breakpointConfig.variant === SmallViewportVariant.centeredoverflow &&
    !canUseCenteredOverflow &&
    breakpointConfig.fallbackVariant === SmallViewportVariant.horizontalscroll
  ) {
    return {
      variant: canUseHorizontalScroll ? SmallViewportVariant.horizontalscroll : SmallViewportVariant.none,
      breakpoint: breakpointConfig.breakpoint,
    };
  }

  if (
    breakpointConfig.variant === SmallViewportVariant.centeredoverflow &&
    !canUseCenteredOverflow &&
    breakpointConfig.fallbackVariant !== SmallViewportVariant.centeredoverflow
  ) {
    return {
      variant: breakpointConfig.fallbackVariant ?? SmallViewportVariant.none,
      breakpoint: breakpointConfig.breakpoint,
    };
  }

  if (
    breakpointConfig.variant === SmallViewportVariant.horizontalscroll &&
    !canUseHorizontalScroll &&
    breakpointConfig.fallbackVariant === SmallViewportVariant.centeredoverflow
  ) {
    return {
      variant: canUseCenteredOverflow ? SmallViewportVariant.centeredoverflow : SmallViewportVariant.none,
      breakpoint: breakpointConfig.breakpoint,
    };
  }

  if (
    breakpointConfig.variant === SmallViewportVariant.horizontalscroll &&
    !canUseHorizontalScroll &&
    breakpointConfig.fallbackVariant !== SmallViewportVariant.horizontalscroll
  ) {
    return {
      variant: breakpointConfig.fallbackVariant ?? SmallViewportVariant.none,
      breakpoint: breakpointConfig.breakpoint,
    };
  }

  return breakpointConfig;
};

/**
 * Finn klassenavn for responsiv oppførsel
 * @param config Konfigurasjon for responsiv oppførsel
 * @returns Klassenavn
 */
export const getBreakpointClass = (config?: BreakpointConfig): string | undefined =>
  config && configUsesCss(config) ? mapConfigToClass(config) : undefined;

/** @deprecated Midlertidig, fjernes i v3.0.0 */
export const isOldFormat = (x: unknown): x is SmallViewportVariant =>
  (x as SmallViewportVariant) === SmallViewportVariant.block || (x as SmallViewportVariant) === SmallViewportVariant.horizontalscroll;

/** @deprecated Midlertidig, fjernes i v3.0.0 */
export const getBackwardsCompatibleConfig = (variant: SmallViewportVariant): BreakpointConfig => {
  return {
    breakpoint: 'sm',
    variant,
  };
};
