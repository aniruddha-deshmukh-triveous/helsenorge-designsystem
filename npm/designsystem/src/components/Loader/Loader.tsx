import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';
import { uuid } from '../../utils/uuid';

import { PaletteNames } from '../../theme/palette';
import loaderStyles from './styles.module.scss';

export type LoaderColors = PaletteNames;
export type LoaderSizes = 'tiny' | 'small' | 'medium' | 'large';
export enum Overlay {
  screen = 'screen',
  parent = 'parent',
}

interface LoaderProps {
  /** Sets the color of the loader */
  color?: LoaderColors;
  /**	Adds custom classes to the element. */
  className?: string;
  /** Changes the size of the loader. */
  size?: LoaderSizes;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Centers the loader in a container */
  center?: boolean;
  /** Loader is displayed with grey background covering the entire screen */
  overlay?: keyof typeof Overlay;
  /**  Individual id for loading icon (aria-labelledby).  */
  ariaLabelledById?: string;
  /**  String that labels the current loading element  */
  ariaLabel?: string;
}

const Loader = React.forwardRef(function LoaderForwardedRef(props: LoaderProps, ref: React.ForwardedRef<HTMLElement>) {
  const {
    overlay,
    color = overlay ? 'black' : 'neutral',
    size = 'small',
    className = '',
    testId,
    center,
    ariaLabelledById,
    ariaLabel = 'Laster inn',
  } = props;

  const [display, setDisplay] = useState(overlay !== Overlay.parent);

  const isSmall = size === 'small';
  const isMedium = size === 'medium';
  const isLarge = size === 'large';

  const loaderWrapperClasses = classNames(loaderStyles['loader-wrapper'], {
    [loaderStyles['loader-wrapper--center']]: center,
    [loaderStyles['loader-wrapper--overlay-screen']]: overlay === Overlay.screen,
    [loaderStyles['loader-wrapper--overlay-parent']]: overlay === Overlay.parent && display,
  });
  const loaderClasses = classNames(
    loaderStyles.loader,
    {
      [loaderStyles['loader--small']]: isSmall,
      [loaderStyles['loader--medium']]: isMedium,
      [loaderStyles['loader--large']]: isLarge,
    },
    className
  );
  const loaderDotClasses = classNames(loaderStyles.loader__dot, {
    [loaderStyles['loader__dot--small']]: isSmall,
    [loaderStyles['loader__dot--medium']]: isMedium,
    [loaderStyles['loader__dot--large']]: isLarge,
    [loaderStyles['loader__dot--banana']]: color === 'banana',
    [loaderStyles['loader__dot--cherry']]: color === 'cherry',
    [loaderStyles['loader__dot--kiwi']]: color === 'kiwi',
    [loaderStyles['loader__dot--neutral']]: color === 'neutral',
    [loaderStyles['loader__dot--plum']]: color === 'plum',
    [loaderStyles['loader__dot--black']]: color === 'black',
    [loaderStyles['loader__dot--white']]: color === 'white',
  });

  const uniqueId = `loader${uuid()}`;

  const referance = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (referance.current?.parentElement?.style && overlay === Overlay.parent) {
      referance.current.parentElement.style.position = 'relative';
      setDisplay(true);
    }
  }, []);

  return (
    <div role="progressbar" className={loaderWrapperClasses} ref={referance}>
      {display && (
        <div data-testid={testId} aria-labelledby={ariaLabelledById || uniqueId} className={loaderClasses}>
          <div className={loaderDotClasses} />
          <div className={loaderDotClasses} />
          <div className={loaderDotClasses} />
          <div className={loaderDotClasses} />
          {!ariaLabelledById && (
            <span id={uniqueId} className={loaderStyles['loader__hidden-text']}>
              {ariaLabel}
            </span>
          )}
        </div>
      )}
    </div>
  );
});

export default Loader;
