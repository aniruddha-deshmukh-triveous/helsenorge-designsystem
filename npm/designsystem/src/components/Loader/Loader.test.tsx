import React from 'react';
import { render, screen } from '@testing-library/react';
import Loader from './Loader';
import * as uuidUtils from '../../utils/uuid';

jest.spyOn(uuidUtils, 'uuid').mockReturnValue(`-unik-id`);

describe('Gitt at en loader skal vises ', (): void => {
  describe('Når at en loader rendres ', (): void => {
    test('Så vises loaderen', (): void => {
      const { container } = render(<Loader />);
      expect(container).toMatchSnapshot();
    });

    test('Så er alle fire loaderDots riktig plassert', (): void => {
      render(<Loader testId={'loaderDotTest'} />);
      const loaderDotsWrapper = screen.getByTestId('loaderDotTest');

      expect(loaderDotsWrapper.children[0].className).toBe('loader__dot loader__dot--small loader__dot--neutral');
      expect(loaderDotsWrapper.children[1].className).toBe('loader__dot loader__dot--small loader__dot--neutral');
      expect(loaderDotsWrapper.children[2].className).toBe('loader__dot loader__dot--small loader__dot--neutral');
      expect(loaderDotsWrapper.children[3].className).toBe('loader__dot loader__dot--small loader__dot--neutral');
    });

    test('Så har loader progressbar role', (): void => {
      render(<Loader overlay />);
      expect(screen.getByRole('progressbar')).toBeVisible();
    });
  });

  describe('Når center prop er true ', (): void => {
    test('Så sentreres loader', (): void => {
      const { container } = render(<Loader center />);
      expect(container.firstChild).toHaveClass('loader-wrapper--center');
    });
  });

  describe('Når overlay prop er satt', (): void => {
    test('Så har loaderen en overlay, og color er nå svart på loaderen', (): void => {
      const { container } = render(<Loader overlay testId={'loaderDotTest'} />);
      expect(container.firstChild).toHaveClass('loader-wrapper--overlay');
      expect(container.firstChild.firstChild).toHaveAttribute('aria-labelledby', 'loader-unik-id');

      const loaderDotsWrapper = screen.getByTestId('loaderDotTest');
      expect(loaderDotsWrapper.children[0].className).toBe('loader__dot loader__dot--small loader__dot--black');
    });
  });

  describe('Når ariaLabelledById er satt', (): void => {
    test('Så settes attributten aria-labelledby til det samme', (): void => {
      const { container } = render(<Loader ariaLabelledById="aria-test" overlay />);
      expect(container.firstChild).toHaveClass('loader-wrapper--overlay');
      expect(container.firstChild.firstChild).toHaveAttribute('aria-labelledby', 'aria-test');
    });
  });

  describe('Når testId-prop er satt', (): void => {
    test('Så kan komponenten finnes ved hjelp av testId', (): void => {
      render(<Loader testId="bare-tester" />);

      const component = screen.getByTestId('bare-tester');
      expect(component).toBeVisible();
    });
  });
});
