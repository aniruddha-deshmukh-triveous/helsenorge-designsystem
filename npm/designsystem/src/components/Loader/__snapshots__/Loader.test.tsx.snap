// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at en loader skal vises  Når at en loader rendres  Så vises loaderen 1`] = `
<div>
  <div
    class="loader-wrapper"
    role="progressbar"
  >
    <div
      aria-labelledby="unik-id"
      class="loader loader--small"
      data-analyticsid="loader"
    >
      <div
        class="loader__dot loader__dot--small loader__dot--neutral"
      />
      <div
        class="loader__dot loader__dot--small loader__dot--neutral"
      />
      <div
        class="loader__dot loader__dot--small loader__dot--neutral"
      />
      <div
        class="loader__dot loader__dot--small loader__dot--neutral"
      />
      <span
        class="loader__hidden-text"
        id="unik-id"
      >
        Laster inn
      </span>
    </div>
  </div>
</div>
`;
